File: benchmark/recursive-tree/tree.rb
class Node
  attr_accessor :value, :left, :right
  
  def initialize(value)
    @value = value
    @left = nil
    @right = nil
  end
  
  def insert(val)
    if val < @value
      if @left.nil?
        @left = Node.new(val)
      else
        @left.insert(val)
      end
    else
      if @right.nil?
        @right = Node.new(val)
      else
        @right.insert(val)
      end
    end
  end
  
  def sum
    total = @value
    total += @left.sum unless @left.nil?
    total += @right.sum unless @right.nil?
    total
  end
  
  def count
    c = 1
    c += @left.count unless @left.nil?
    c += @right.count unless @right.nil?
    c
  end
  
  def height
    lh = @left.nil? ? 0 : @left.height
    rh = @right.nil? ? 0 : @right.height
    [lh, rh].max + 1
  end
end

def make_tree(values)
  return nil if values.empty?
  root = Node.new(values[0])
  values[1..-1].each { |val| root.insert(val) }
  root
end

def fibonacci(n)
  return n if n <= 1
  fibonacci(n - 1) + fibonacci(n - 2)
end

def process_with_closure(arr, multiplier)
  fib10 = fibonacci(10)
  
  transform = ->(x) { x * multiplier + fib10 }
  
  arr.map { |x| transform.call(x) }
end

def matrix_multiply(a, b)
  n = a.length
  result = Array.new(n) { Array.new(n, 0) }
  
  n.times do |i|
    n.times do |j|
      sum = 0
      n.times do |k|
        sum += a[i][k] * b[k][j]
      end
      result[i][j] = sum
    end
  end
  
  result
end

def main
  iterations = 500
  tree_size = 100
  matrix_size = 20
  
  total_sum = 0
  total_count = 0
  total_height = 0
  
  iterations.times do |iter|
    values = (0...tree_size).map { |i| (i * 7 + iter * 13) % 1000 }
    
    tree = make_tree(values)
    total_sum += tree.sum
    total_count += tree.count
    total_height += tree.height
    
    processed = process_with_closure(values, iter % 10 + 1)
    processed.each { |val| total_sum += val % 100 }
    
    if iter % 50 == 0
      mat_a = Array.new(matrix_size) { |i| Array.new(matrix_size) { |j| (i + j + iter) % 10 } }
      mat_b = Array.new(matrix_size) { |i| Array.new(matrix_size) { |j| (i * j + iter) % 10 } }
      
      mat_c = matrix_multiply(mat_a, mat_b)
      mat_c.each do |row|
        row.each { |val| total_sum += val }
      end
    end
  end
  
  puts "#{total_sum} #{total_count} #{total_height}"
end

main

File: benchmark/recursive-tree/Tree.java
class Node {
    int value;
    Node left;
    Node right;
    
    Node(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
    
    void insert(int val) {
        if (val < this.value) {
            if (this.left == null) {
                this.left = new Node(val);
            } else {
                this.left.insert(val);
            }
        } else {
            if (this.right == null) {
                this.right = new Node(val);
            } else {
                this.right.insert(val);
            }
        }
    }
    
    int sum() {
        int total = this.value;
        if (this.left != null) {
            total += this.left.sum();
        }
        if (this.right != null) {
            total += this.right.sum();
        }
        return total;
    }
    
    int count() {
        int c = 1;
        if (this.left != null) {
            c += this.left.count();
        }
        if (this.right != null) {
            c += this.right.count();
        }
        return c;
    }
    
    int height() {
        int lh = 0;
        int rh = 0;
        if (this.left != null) {
            lh = this.left.height();
        }
        if (this.right != null) {
            rh = this.right.height();
        }
        return Math.max(lh, rh) + 1;
    }
}

public class Tree {
    static Node makeTree(int[] values) {
        if (values.length == 0) {
            return null;
        }
        Node root = new Node(values[0]);
        for (int i = 1; i < values.length; i++) {
            root.insert(values[i]);
        }
        return root;
    }
    
    static int fibonacci(int n) {
        if (n <= 1) {
            return n;
        }
        return fibonacci(n - 1) + fibonacci(n - 2);
    }
    
    static int[] processWithClosure(int[] arr, int multiplier) {
        int[] result = new int[arr.length];
        int fib10 = fibonacci(10);
        
        for (int i = 0; i < arr.length; i++) {
            result[i] = arr[i] * multiplier + fib10;
        }
        
        return result;
    }
    
    static int[][] matrixMultiply(int[][] a, int[][] b) {
        int n = a.length;
        int[][] result = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                int sum = 0;
                for (int k = 0; k < n; k++) {
                    sum += a[i][k] * b[k][j];
                }
                result[i][j] = sum;
            }
        }
        
        return result;
    }
    
    public static void main(String[] args) {
        int iterations = 500;
        int treeSize = 100;
        int matrixSize = 20;
        
        int totalSum = 0;
        int totalCount = 0;
        int totalHeight = 0;
        
        for (int iter = 0; iter < iterations; iter++) {
            int[] values = new int[treeSize];
            for (int i = 0; i < treeSize; i++) {
                values[i] = (i * 7 + iter * 13) % 1000;
            }
            
            Node tree = makeTree(values);
            totalSum += tree.sum();
            totalCount += tree.count();
            totalHeight += tree.height();
            
            int[] processed = processWithClosure(values, iter % 10 + 1);
            for (int i = 0; i < processed.length; i++) {
                totalSum += processed[i] % 100;
            }
            
            if (iter % 50 == 0) {
                int[][] matA = new int[matrixSize][matrixSize];
                int[][] matB = new int[matrixSize][matrixSize];
                
                for (int i = 0; i < matrixSize; i++) {
                    for (int j = 0; j < matrixSize; j++) {
                        matA[i][j] = (i + j + iter) % 10;
                        matB[i][j] = (i * j + iter) % 10;
                    }
                }
                
                int[][] matC = matrixMultiply(matA, matB);
                for (int i = 0; i < matrixSize; i++) {
                    for (int j = 0; j < matrixSize; j++) {
                        totalSum += matC[i][j];
                    }
                }
            }
        }
        
        System.out.println(totalSum + " " + totalCount + " " + totalHeight);
    }
}
